internal/proto/examples/scalars_canoto_test.go

```diff
@@ -0,0 +1,314 @@
+				Sint64:   s64,
+				Fixed32:  f32,
+				Fixed64:  f64,
+				Sfixed32: sf32,
+				Sfixed64: sf64,
+				Bool:     b,
+				String_:  s,
+				Bytes:    bs,
+			},
+			pb.Scalars{
+				Int32:    pbScalars.Int32,
+				Int64:    pbScalars.Int64,
+				Uint32:   pbScalars.Uint32,
+				Uint64:   pbScalars.Uint64,
+				Sint32:   pbScalars.Sint32,
+				Sint64:   pbScalars.Sint64,
+				Fixed32:  pbScalars.Fixed32,
+				Fixed64:  pbScalars.Fixed64,
+				Sfixed32: pbScalars.Sfixed32,
+				Sfixed64: pbScalars.Sfixed64,
+				Bool:     pbScalars.Bool,
+				String_:  pbScalars.String_,
+				Bytes:    pbScalars.Bytes,
+			},
+		)
+	})
+}
+
+func FuzzScalars_Canonical(f *testing.F) {
+	f.Fuzz(func(t *testing.T, b []byte) {
+		require := require.New(t)
+
+		var (
+			scalars Scalars
+			reader  = canoto.Reader{B: b}
+		)
+		err := scalars.UnmarshalCanoto(&reader)
+		if err != nil {
+			return
+		}
+
+		size := scalars.SizeCanoto()
+		require.Len(b, size)
+
+		w := canoto.Writer{
+			B: make([]byte, 0, size),
+		}
+		scalars.MarshalCanoto(&w)
+		require.Equal(b, w.B)
+	})
+}
+
+func BenchmarkScalars_MarshalCanoto(b *testing.B) {
+	var (
+		s     = "asdkcjhbsakjdhcbasdc"
+		bytes = []byte("kjsadhbcaskjdvhb")
+	)
+	for range b.N {
+		cbScalars := Scalars{
+			Int32:    216457,
+			Int64:    -2138746,
+			Uint32:   32485976,
+			Uint64:   287634,
+			Sint32:   -12786345,
+			Sint64:   98761243,
+			Fixed32:  98765234,
+			Fixed64:  1234576,
+			Sfixed32: -21348976,
+			Sfixed64: 98756432,
+			Bool:     true,
+			String:   s,
+			Bytes:    bytes,
+		}
+
+		size := cbScalars.SizeCanoto()
+		w := canoto.Writer{
+			B: make([]byte, 0, size),
+		}
+		cbScalars.MarshalCanoto(&w)
+	}
+}
+
+func BenchmarkScalars_UnmarshalCanoto(b *testing.B) {
+	cbScalars := Scalars{
+		Int32:    216457,
+		Int64:    -2138746,
+		Uint32:   32485976,
+		Uint64:   287634,
+		Sint32:   -12786345,
+		Sint64:   98761243,
+		Fixed32:  98765234,
+		Fixed64:  1234576,
+		Sfixed32: -21348976,
+		Sfixed64: 98756432,
+		Bool:     true,
+		String:   "asdkcjhbsakjdhcbasdc",
+		Bytes:    []byte("kjsadhbcaskjdvhb"),
+	}
+	w := canoto.Writer{}
+	cbScalars.MarshalCanoto(&w)
+
+	for _, unsafe := range []bool{false, true} {
+		b.Run("unsafe="+strconv.FormatBool(unsafe), func(b *testing.B) {
+			for range b.N {
+				var (
+					scalars Scalars
+					reader  = canoto.Reader{
+						B:      w.B,
+						Unsafe: unsafe,
+					}
+				)
+				_ = scalars.UnmarshalCanoto(&reader)
+			}
+		})
+	}
+}
+
+func BenchmarkScalars_MarshalProto(b *testing.B) {
+	var (
+		s     = "asdkcjhbsakjdhcbasdc"
+		bytes = []byte("kjsadhbcaskjdvhb")
+	)
+	for range b.N {
+		pbScalars := pb.Scalars{
+			Int32:    216457,
+			Int64:    -2138746,
+			Uint32:   32485976,
+			Uint64:   287634,
+			Sint32:   -12786345,
+			Sint64:   98761243,
+			Fixed32:  98765234,
+			Fixed64:  1234576,
+			Sfixed32: -21348976,
+			Sfixed64: 98756432,
+			Bool:     true,
+			String_:  s,
+			Bytes:    bytes,
+		}
+		_, _ = proto.Marshal(&pbScalars)
```

Comment:
This benchmark is setup to showcase that passing in the struct like this causes the struct to be allocated on the heap. While that is probably going to be the case for most usages... If someone actually cares about performance, it is really nice to provide a 0 memory allocation API